### ISSUE TITLE
Dangerous workflows: detect refs

### ISSUE THREAD

**Issue by laurentsimon** [ossf/scorecard#2404]
Workflows that use labels to enforce reviews before running test on pull_request_targets checkout the repo with `ref: ${{github.event.pull_request.head.sha}}`. One mistake some developers make is use `ref: ${{github.event.pull_request.head.ref}}` instead, which is subject to a TOCTOU attack.

1. Maintainer reviews the RP code and adds a label
2. Attacker pushes different code before the workflow is triggered / the repo is checked out.

Scorecard can detect these as part of the dangerous workflow check.

**Comment by spencerschrock** [comment]
Scorecard currently detects both patterns as part of Dangerous workflows. In fact, we're currently getting dinged with it:

```json
"checks": [
    {
      "details": [
        "Warn: untrusted code checkout '${{ github.event.pull_request.head.sha }}': .github/workflows/integration.yml:45"
      ],
      "score": 0,
      "reason": "dangerous workflow patterns detected",
      "name": "Dangerous-Workflow",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/main/docs/checks.md#dangerous-workflow",
        "short": "Determines if the project's GitHub Action workflows avoid dangerous patterns."
      }
    }
  ],
  ```
  
  I assume the intended behavior should be?
  not dangerous: `github.event.pull_request.head.sha`
  dangerous: `github.event.pull_request.head.ref`

**Comment by laurentsimon** [comment]
yes you're correct @spencerschrock I think we do catch it today. `sha` is secure if used the code is not run afterwards or if labels / env are applied prior to run. I'm not 100% sure if we can capture all the distinction, though. We already detect the use of env, but not labels, for example.

**Comment by kyzzniko-lang** [comment]
I found a related Dangerous-Workflow false negative involving an explicit fork checkout. On a pull_request_target workflow, the following checkout is attacker-controlled but is not currently reported:

``yaml
- uses: actions/checkout@v4
  with:
    repository: ${{ github.event.pull_request.head.repo.full_name }}
    ref: ${{ github.head_ref }}
    allow-unsafe-pr-checkout: true
``

The current check only recognizes ref values containing github.event.pull_request or github.event.workflow_run; it does not combine the attacker-controlled repository input with github.head_ref. I reproduced this with a focused regression test: the existing head.sha fixture is detected, while this fixture produces zero findings. Would a narrowly scoped PR adding this detection and safe negative cases be welcome?
